interceptors_node_publicist.go

```diff
@@ -0,0 +1,397 @@
+		ctx, req.VolumeId, req.PublishVolumeInfo, req.UserCredentials)
+	if err != nil {
+		return err
+	}
+	if devicePath == "" {
+		return status.Errorf(
+			codes.FailedPrecondition,
+			"failed to discover device path for volume: %s", req.VolumeId)
+	}
+	if i.validateDevice {
+		validatedDevPath, err := validateDevice(devicePath)
+		if err != nil {
+			return status.Error(codes.FailedPrecondition, err.Error())
+		}
+		devicePath = validatedDevPath
+	}
+	lf["devicePath"] = devicePath
+
+	privMntTgtName, err := i.getPrivateMountTargetName(
+		ctx, req.VolumeId, req.UserCredentials)
+	if err != nil {
+		return err
+	}
+	privMntTgtPath := path.Join(i.privateMountDir, privMntTgtName)
+	lf["privateMountTargetPath"] = privMntTgtPath
+
+	mountTable, err := gofsutil.GetMounts(ctx)
+	if err != nil {
+		return err
+	}
+
+	var (
+		isPub     bool
+		isPrivPub bool
+	)
+
+	for _, m := range mountTable {
+		// Check to see if the device is already mounted to the private
+		// target path.
+		if m.Source == devicePath && m.Path == privMntTgtPath {
+			stat, _ := os.Stat(privMntTgtPath)
+			if stat == nil {
+				continue
+			}
+			if (isBlock && stat.IsDir()) || !stat.IsDir() {
+				return fmt.Errorf(
+					"invalid existing private mount target: %s",
+					privMntTgtPath)
+			}
+			isPrivPub = true
+			lf["deviceToPrivate"] = isPrivPub
+		}
+
+		// Check to see if the private target path is already mounted to the
+		// target path.
+		if m.Source == privMntTgtPath && m.Path == req.TargetPath {
+			stat, _ := os.Stat(req.TargetPath)
+			if stat == nil {
+				continue
+			}
+			if (isBlock && stat.IsDir()) || !stat.IsDir() {
+				return fmt.Errorf(
+					"invalid existing target path: %s",
+					req.TargetPath)
+			}
+			isPub = true
+			lf["privateToPublic"] = isPub
+		}
+	}
+
+	if !isPrivPub {
+		if isBlock {
+			st, err := os.Stat(privMntTgtPath)
+			if err != nil {
+				if os.IsNotExist(err) {
+					log.WithFields(lf).Debug(
+						"creating private mount target file")
+					f, err := os.Create(privMntTgtPath)
+					if err != nil {
+						return err
+					}
+					f.Close()
+				} else {
+					return err
+				}
+			}
+			log.WithFields(lf).Debug(
+				"binding mounting device to private mount target")
+
+			err := gofsutil.BindMount(ctx, devicePath, privMntTgtPath)
+			if err != nil {
+				return err
+			}
+		} else {
+			st, err := os.Stat(privMntTgtPath)
+			if err != nil {
+				if os.IsNotExist(err) {
+					log.WithFields(lf).Debug(
+						"creating private mount target dir")
+					if err := os.MkdirAll(privMntTgtPath); err != nil {
+						return err
+					}
+				} else {
+					return err
+				}
+			}
+			log.WithFields(lf).Debug(
+				"formatting device & mounting to private mount target")
+			if f := i.p.FormatAndMount(); f != nil {
+				err := f(
+					ctx, devicePath, req.TargetPath, fsType,
+					mntFlags, req.UserCredentials)
+				if err != nil {
+					return err
+				}
+			} else {
+				err := gofsutil.FormatAndMount(
```

Comment:
I know in the previous code, we only allowed a call to `format and mount` if we weren't mounting something readonly. That way, even if it was a brand new volume, if the access mode was readonly, we woudln't format it. There's no point  in formatting a brand new volume if the mode is going to be read-only - there won't be anything there.